File: Projekti/SQLAlchemy-Benchmark/repository/tag_repository.py
from model.tag import Tag
from model.user import User
from model.tag import Tag
from model.post import Post, post_tags_table
from sqlalchemy import func, text


class TagRepository:

    def __init__(self, session):
        self.session = session

    def get_tags_by_user(self, email):
        return self.session.query(Tag.tag_name)\
            .distinct()\
            .join(post_tags_table, Tag.tag_id == post_tags_table.c.tag_id)\
            .join(Post, Post.post_id == post_tags_table.c.post_id)\
            .join(User, User.user_id == Post.user_id)\
            .filter(User.email == email).all()

    def count_posts_per_tag(self):
        return self.session.query(Tag.tag_name, func.count(Post.post_id).label('total_posts'))\
            .join(post_tags_table, Tag.tag_id == post_tags_table.c.tag_id)\
            .join(Post, Post.post_id == post_tags_table.c.post_id)\
            .group_by(Tag.tag_name)\
            .having(text('total_posts >= 5')).all()

    def count_tags_per_user(self):
        query = self.session.query(Post.user_id, func.count(post_tags_table.c.tag_id)
                                   .label('total_tags'))\
            .join(post_tags_table, Post.post_id == post_tags_table.c.post_id)\
            .group_by(Post.user_id)
        print(query)
        return query.all()

File: Projekti/Hibernate-Benchmark/src/main/java/com/example/HibernateBenchmark/repository/TagRepository.java
package com.example.HibernateBenchmark.repository;

import com.example.HibernateBenchmark.dto.TagCount;
import com.example.HibernateBenchmark.model.Tag;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;

public interface TagRepository extends JpaRepository<Tag, Integer> {

    @Query("SELECT DISTINCT t.name FROM User u JOIN u.posts p JOIN p.tags t WHERE u.email = :email")
    List<String> findDistinctTagsByEmail(@Param("email") String email);

    @Query("SELECT t.name, COUNT(p.postId) FROM Post p JOIN p.tags t GROUP BY t.name HAVING COUNT(p.postId) >= 5")
    List<TagCount> countPostsPerTag();

    @Query("SELECT p.user.userId, COUNT(t.tagId) FROM Post p JOIN p.tags t GROUP BY p.user")
    List<Object[]> countTagsPerUser();
}
